### PR TITLE
feat: new transfer transaction

### PR DESCRIPTION
**Pull request by c0nsol3**
Transaction: https://texplorer.solar.org/transactions/a1a8eb81d92d4ab0d284a8e81abdca2333161f5a36e812f5461b8469f55e1d10

Related: https://github.com/Solar-network/python-crypto/issues/31

### CODE REVIEW

File: solar_crypto/constants.py [hunk 1]
```diff
@@ -1,25 +1,25 @@
...
 
-TRANSACTION_TRANSFER = 0
+TRANSACTION_LEGACY_TRANSFER = 0
 TRANSACTION_SECOND_SIGNATURE_REGISTRATION = 1
 TRANSACTION_DELEGATE_REGISTRATION = 2
 TRANSACTION_VOTE = 3
 TRANSACTION_MULTI_SIGNATURE_REGISTRATION = 4
 TRANSACTION_IPFS = 5
-TRANSACTION_MULTI_PAYMENT = 6
+TRANSFER = 6
```
Comment (dutchpool) [inline solar_crypto/constants.py:9]: Should be called `TRANSACTION_TRANSFER` right?

Comment (c0nsol3) [reply solar_crypto/constants.py:9]: Yes, will fix. Thanks

File: solar_crypto/transactions/builder/transfer.py [hunk 2]
```diff
@@ -7,28 +7,33 @@ class Transfer(BaseTransactionBuilder):
 
     transaction_type = TRANSACTION_TRANSFER
 
-    def __init__(self, recipientId, amount, vendorField=None, fee=None):
-        """Create a transfer transaction
+    def __init__(self, vendorField=None, fee=None):
+        """Create a multi payment transaction
```
Comment (sleepdefic1t) [inline solar_crypto/transactions/builder/transfer.py:11]:
```diff
- Create a multi payment transaction
+ Create a transfer transaction
```

Comment (sleepdefic1t) [reply solar_crypto/transactions/builder/transfer.py:11]: The term "multi payment" is no longer being used.

File: solar_crypto/transactions/serializers/transfer.py [hunk 3]
```diff
@@ -2,23 +2,25 @@
...
+from binary.unsigned_integer.writer import write_bit16, write_bit64
 
 from solar_crypto.exceptions import SolarSerializerException
 from solar_crypto.identity import address
 from solar_crypto.transactions.serializers.base import BaseSerializer
 
 
 class TransferSerializer(BaseSerializer):
-    """Serializer handling transfer data"""
+    """Serializer handling multi payment data"""
```
Comment (sleepdefic1t) [inline solar_crypto/transactions/serializers/transfer.py:13]:
```diff
- """Serializer handling multi payment data"""
+ """Serializer handling transfer data"""
```

File: tests/transactions/builder/test_transfer.py [hunk 4]
```diff
@@ -1,176 +1,127 @@
...
 
 
 @pytest.mark.parametrize("version", [2, 3])
-def test_transfer_transaction_custom_fee_via_kwargs(version):
-    """Test if a transfer transaction gets built with a custom fee"""
-    transaction = Transfer(
-        recipientId="D61mfSggzbvQgTUe6JhYKH2doHaqJ3Dyib", amount=200000000, fee=5
-    )
+def test_transaction_custom_fee_via_method(version):
+    """Test if multi payment transaction gets built with a custom fee"""
```
Comment (sleepdefic1t) [inline tests/transactions/builder/test_transfer.py:75]:
```diff
- """Test if multi payment transaction gets built with a custom fee"""
+ """Test if transfer transaction gets built with a custom fee"""
```

File: tests/transactions/builder/test_transfer.py [hunk 5]
```diff
@@ -1,176 +1,127 @@
...
     transaction.verify()  # if no exception is raised, it means the transaction is valid
 
 
 @pytest.mark.parametrize("version", [2, 3])
-def test_transfer_transaction_update_amount(version):
-    """Test if a transfer transaction can update an amount"""
-    transaction = Transfer(recipientId="D61mfSggzbvQgTUe6JhYKH2doHaqJ3Dyib", amount=200000000)
-    transaction.set_amount(10)
+def test_transaction_custom_fee_via_kwargs(version):
+    """Test if multi payment transaction gets built with a custom fee"""
```
Comment (sleepdefic1t) [inline tests/transactions/builder/test_transfer.py:44]:
```diff
- """Test if multi payment transaction gets built with a custom fee"""
+ """Test if transfer transaction gets built with a custom fee"""
```

File: tests/transactions/builder/test_transfer.py [hunk 6]
```diff
@@ -1,176 +1,127 @@
...
 
 @pytest.mark.parametrize("version", [2, 3])
-def test_transfer_transaction(version):
-    """Test if a transfer transaction gets built"""
-    transaction = Transfer(
-        recipientId="D61mfSggzbvQgTUe6JhYKH2doHaqJ3Dyib",
-        amount=200000000,
-    )
+def test_transaction(version):
+    """Test if multi payment transaction gets built"""
```
Comment (sleepdefic1t) [inline tests/transactions/builder/test_transfer.py:13]:
```diff
- """Test if multi payment transaction gets built"""
+ """Test if transfer transaction gets built"""
```

File: tests/conftest.py [hunk 7]
```diff
@@ -146,7 +146,7 @@ def transaction_type_6():
     data = {
```
Comment (sleepdefic1t) [inline tests/conftest.py:146]: [#L145](https://github.com/c0nsol3/python-crypto-solar/blob/multitransfer/tests/conftest.py#L145)

```diff
- """Transaction of type "multi payment" """
+ """Transaction of type "transfer" """
```

**Review by sleepdefic1t — CHANGES_REQUESTED** [review]
This looks really good.
The only changes needed here involve comments:
- `multi payment` => `transfer`

File: solar_crypto/transactions/deserializers/transfer.py [hunk 8]
```diff
@@ -10,15 +10,30 @@ class TransferDeserializer(BaseDeserializer):
     def deserialize(self):
         starting_position = int(self.asset_offset / 2)
 
-        self.transaction.amount = read_bit64(self.serialized, offset=starting_position)
-        self.transaction.expiration = read_bit32(self.serialized, offset=starting_position + 8)
+        payment_length = read_bit16(self.serialized, starting_position) & 0xFF
```
Comment (sleepdefic1t) [inline solar_crypto/transactions/deserializers/transfer.py:13]:
```diff
- payment_length
+ transfers_length
```

**Review by sleepdefic1t — CHANGES_REQUESTED** [review]

File: solar_crypto/transactions/deserializers/transfer.py (hunk 8 above)
Comment (sleepdefic1t) [reply solar_crypto/transactions/deserializers/transfer.py:13]: #ln19 and 36 will need to reflect this change too 👍

File: solar_crypto/transactions/builder/transfer.py [hunk 9]
```diff
@@ -7,28 +7,33 @@ class Transfer(BaseTransactionBuilder):
...
 
         self.transaction.vendorField = vendorField.encode() if vendorField else None
 
         if fee:
             self.transaction.fee = fee
+
+    def get_type_group(self):
+        return TRANSACTION_TYPE_GROUP.CORE.value
+
+    def add_payment(self, amount, recipient_id):
```
Comment (sleepdefic1t) [inline solar_crypto/transactions/builder/transfer.py:31]: Is there any sort of aliasing/backwards compatibility where this (`def add_payment(self, amount, recipient_id):`) shouldn't just be removed altogether?

File: solar_crypto/transactions/serializers/transfer.py [hunk 10]
```diff
@@ -13,12 +13,14 @@ class TransferSerializer(BaseSerializer):
...
     def serialize(self):
-        if not address.validate_address(self.transaction["recipientId"]):
-            raise SolarSerializerException("Invalid recipient address")
+        self.bytes_data += write_bit16(len(self.transaction["asset"]["transfers"]))
 
-        self.bytes_data += write_bit64(self.transaction["amount"])
-        self.bytes_data += write_bit32(self.transaction.get("expiration", 0))
-        recipientId = hexlify(b58decode_check(self.transaction["recipientId"]))
-        self.bytes_data += write_high(recipientId)
+        for payment in self.transaction["asset"]["transfers"]:
```
Comment (sleepdefic1t) [inline solar_crypto/transactions/serializers/transfer.py:18]:
```diff
- for payment in self.transaction["asset"]["transfers"]:
+ for transfer in self.transaction["asset"]["transfers"]:
```

**Review by sleepdefic1t — CHANGES_REQUESTED** [review]

File: solar_crypto/transactions/builder/transfer.py (hunk 9 above)
Comment (c0nsol3) [reply solar_crypto/transactions/builder/transfer.py:31]: Only reason why I added it was to have the same methods as core. I'll remove it.

**Comment by c0nsol3** [comment]
I've made the adjustment in my last commit.

File: solar_crypto/transactions/builder/transfer.py (hunk 9 above)
Comment (sleepdefic1t) [reply solar_crypto/transactions/builder/transfer.py]: Right on. TS allows 'multipayment' to be an alias for 'transfer' (i.e., backwards compatibility). If that's not a thing here, it's not needed 👍

Thank you, @c0nsol3 🙏

**Pull request merged**